### PR TITLE
prov/efa: fix getinfo path to validate FI_HMEM/FI_MR_HMEM

### PR DESCRIPTION
**Pull request by rwespetal**
When the FI_HMEM capabability is set FI_MR_HMEM is required.

### CODE REVIEW

File: prov/efa/src/rxr/rxr_init.c [hunk 1]
```diff
@@ -381,6 +381,10 @@ static int rxr_info_to_rxr(uint32_t version, const struct fi_info *core_info,
 		 * which means FI_MR_HMEM implies FI_MR_LOCAL for cuda buffer
 		 */
 		if (hints->caps & FI_HMEM) {
+			if (hints->domain_attr &&
+			    !(hints->domain_attr->mr_mode & FI_MR_HMEM))
```
Comment (wzamazon) [inline prov/efa/src/rxr/rxr_init.c:385]: Maybe add a warning here?

File: prov/efa/src/rxr/rxr_init.c [hunk 2]
```diff
@@ -591,6 +592,12 @@ static int rxr_getinfo(uint32_t version, const char *node,
 			assert(!strcmp(shm_info->fabric_attr->name, "shm"));
 		}
 	}
+
+	goto out;
```
Comment (wzamazon) [inline prov/efa/src/rxr/rxr_init.c:596]: This is the success path, can we just do
 ```fi_freeinfo(core_info); return 0``` here 
instead ```goto```?

 Mixing success path and error handling path makes me a little nervous.

File: prov/efa/src/rxr/rxr_init.c (hunk 1 above)
Comment (rwespetal) [reply prov/efa/src/rxr/rxr_init.c]: Sure, added.

File: prov/efa/src/rxr/rxr_init.c (hunk 2 above)
Comment (rwespetal) [reply prov/efa/src/rxr/rxr_init.c]: Good point, fixed.

File: prov/efa/src/rxr/rxr_init.c [hunk 3]
```diff
@@ -381,6 +381,13 @@ static int rxr_info_to_rxr(uint32_t version, const struct fi_info *core_info,
 		 * which means FI_MR_HMEM implies FI_MR_LOCAL for cuda buffer
 		 */
 		if (hints->caps & FI_HMEM) {
+			if (hints->domain_attr &&
+			    !(hints->domain_attr->mr_mode & FI_MR_HMEM)) {
+				FI_INFO(&rxr_prov, FI_LOG_CORE,
```
Comment (wzamazon) [inline prov/efa/src/rxr/rxr_init.c:386]: FI_WARN?

Comment (rwespetal) [reply prov/efa/src/rxr/rxr_init.c:386]: I was following the other validate prints in the getinfo path that use FI_INFO, but I can change this if you want.

Comment (wzamazon) [reply prov/efa/src/rxr/rxr_init.c:386]: If other validation is also using FI_INFO then no need to change.

File: prov/efa/src/rxr/rxr_init.c [hunk 4]
```diff
@@ -591,9 +595,14 @@ static int rxr_getinfo(uint32_t version, const char *node,
...
 		}
 	}
-out:
+
 	fi_freeinfo(core_info);
 	return ret;
+free_info:
+	fi_freeinfo(util_info);
+	fi_freeinfo(*info);
+	*info = NULL;
```
Comment (wzamazon) [inline prov/efa/src/rxr/rxr_init.c:604]: do we need ```free(core_info)``` here?

Comment (rwespetal) [reply prov/efa/src/rxr/rxr_init.c:605]: Yep, you're right, fixed. Looks like that was missing previously in the error paths.

**Pull request merged**